### ISSUE TITLE
Rails 6 inconsistently overrides ActiveJob queue_adapter setting with TestAdapter

### ISSUE THREAD

**Issue by searls** [rails/rails#37270]
### Summary

Like many folks, I set this in my `config/environments/test.rb` file:

```ruby
config.active_job.queue_adapter = :inline  
```

The reason I do this is to have easy-to-test, deterministic behavior of any potentially async delayed jobs triggered by the application. Under Rails 5, all of my tests (children of `ActiveSupport::TestCase` and `ActionDispatch::SystemTestCase` alike) would respect this queue adapter. However, under Rails 6, any built-in Rails test case that includes `ActiveJob::TestHelper` will also override this setting for all descendants of `ActiveJob::Base` ([relevant source](https://github.com/rails/rails/blob/master/activejob/lib/active_job/test_helper.rb#L39-L41)).

The net effect of this is that my `perform_later` and `deliver_later` calls to jobs aren't being fired synchronously (or, in fact, at all) during my system tests, which is leading to test failures upon upgrading to Rails 6.

### Steps to reproduce

Here are two tests that demonstrates the problem:

```ruby
require "test_helper"

class JobTest < ActiveSupport::TestCase
  class SomeJob < ActiveJob::Base
    cattr_accessor :job_ran

    def perform
      @@job_ran = true
    end
  end

  def test_some_job
    assert_equal :inline, Rails.application.config.active_job.queue_adapter
    assert_equal ActiveJob::QueueAdapters::InlineAdapter, SomeJob.queue_adapter.class

    SomeJob.perform_later

    assert_equal true, SomeJob.job_ran
  end
end

class JobSystemTest < ActionDispatch::SystemTestCase
  class OtherJob < ActiveJob::Base
    cattr_accessor :job_ran

    def perform
      @@job_ran = true
    end
  end

  def test_other_job
    assert_equal :inline, Rails.application.config.active_job.queue_adapter
    assert_equal ActiveJob::QueueAdapters::InlineAdapter, OtherJob.queue_adapter.class

    OtherJob.perform_later

    assert_equal true, OtherJob.job_ran
  end
end
```

### Expected behavior

I would expect the behavior to have remained as it did under Rails 5, where the `active_job.queue_adapter` setting would be respected for all of Rails' test types.

### Actual behavior

The first test, which descends from `ActiveSupport::TestCase` will pass and behave as it did under Rails 5. However, the second test, descending from from `ActionDispatch::SystemTestCase` will now fail under Rails 6, because `OtherJob`'s `queue_adapter` will have been reset by `ActiveJob::TestHelper` to be an instance of `TestAdapter`

```
JobSystemTest#test_other_job [/Users/justin/code/testdouble/present/test/lib/job_test.rb:33]:
--- expected
+++ actual
@@ -1 +1 @@
-ActiveJob::QueueAdapters::InlineAdapter
+ActiveJob::QueueAdapters::TestAdapter
```

### System configuration
**Rails version**: 6.0.0 as well as [6-0-stable](https://github.com/rails/rails/tree/2bc6a948a2f46f8cf2d44506af1be235691c84b5), as of today.

**Ruby version**: ruby 2.6.3p62 (2019-04-16 revision 67580) [x86_64-darwin18]

**Comment by searls** [comment]
By the way, for anyone looking to just disable this, one might write a setup hook like this one, invoking the `disable_test_adapter` method also added by the [ActiveJob::TestHelper](https://github.com/rails/rails/blob/master/activejob/lib/active_job/test_helper.rb#L24-L26):

```ruby
  def setup
    (ActiveJob::Base.descendants << ActiveJob::Base).each do |job_class|
      job_class.disable_test_adapter
    end
  ensure
    super
  end
```

**Comment by gmcgibbon** [comment]
I'm slightly confused by this, I see the [same adapter overriding code on 5.2](https://github.com/rails/rails/blob/5-2-stable/activejob/lib/active_job/test_helper.rb#L40-L42). How is it that the behaviour of Rails 5 is different?

We could move this code to the assertion, but that might be expensive. I'm curious why you need to include `ActiveJob::TestHelper` if you aren't planning on using it. If you need both, it could be that the best answer is overriding `queue_adapter_for_test`.

**Comment by searls** [comment]
@gmcgibbon the difference is that my app's test code isn't doing anything to include `ActiveJob::TestHelper`, it's getting invoked anyway, and that's the change I saw when upgrading to Rails 6.

**Comment by gmcgibbon** [comment]
🤦‍♂ My mistake, sorry about that. Yes, the problem was introduced in https://github.com/rails/rails/pull/33849. However, this is fixed on master by https://github.com/rails/rails/pull/36283. 

I think we need to backport https://github.com/rails/rails/pull/36283 + https://github.com/rails/rails/pull/36741 to `6-0-stable` to patch this up.

**Comment by gmcgibbon** [comment]
Backported in ea77dbffdc81d6e40bbde155961ab57f1450d1e4, 1a09b6ebfb75a26dbb4143bf220733aefa854dd8, and 109aad7ca93f0bcb2319df19b45f05b6794b3b92. Let me know if you're still having issues against `6-0-stable` and we can reopen. Thanks!

**Comment by halo** [comment]
**TL; DR:** Upgrading from Rails 5 to 6 caused this bug. However, I cannot confirm that `6-0-stable` or `master` fixed the problem.

### Background

When I enqueue a job via Ajax, I'm using the `provider_job_id` to set up a websocket listener to provide visual feedback on how this particular job is doing.

For some reasone, the `provider_job_id` is not properly kept between requests when I use the `TestAdapter` (not sure why, I think it's simply not persisting it). So I've been using Sidekiq even in tests:

```ruby
# config/application.rb
ActiveJob::Base.queue_adapter = :sidekiq
```

But I have Sidekiq run the jobs inline in feature tests (that way, the `provider_job_id` is reliably kept):

```ruby
# spec/rails_helper.rb
RSpec.configure do |config|
  config.around sidekiq: true do |example|
    Sidekiq::Testing.inline! { example.run }
  end
end
```

So, my capybara tests started failing as of Rails 6:

```ruby
RSpec.describe 'Worksheet', :js, :sidekiq, type: :system do
  it 'can Preview PDFs' do
    document = create :document
    visit document_path document.id

    click_on 'PDF Preview'

    p ActiveJob::Base.queue_adapter  # <- ActiveJob::QueueAdapters::TestAdapter
    
    # This test passes if I uncommend the following line:
    # (ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter)

    p ActiveJob::Base.queue_adapter  # <- Becomes Sidekiq only when disabling the test adapter

    click_on 'Download generated PDF'
  end
end
```

I could not even move the patch to the RSpec `config.around` block, it works only right here, in the test.

I started noticing in my test logs that, in Rails 6, the job is only enqueued but not run. (And when I wrapped the test in `perform_enqueued_jobs do`, the job ran, but the `provider_job_id` was not matching from when the job was enqueued compared to when the websocket channel is set up.)

Thank you @searls for the detailed bug report, can you confirm whether `6-0-stable` fixed this for you?

@gmcgibbon Did you try the "Steps to reproduce" from the issue description on `6-0-master`? I'll try that.

I'm truly sorry I wasn't there when the Rails 6 release candidates were battle tested 😢 Let me know if I can be of further assistance.

**Comment by searls** [comment]
Thanks for the suggestion, @halo 

I just pulled down 6-0-0-stable and can confirm it does not fix this in my case, either

**Comment by gmcgibbon** [comment]
🤔 I'm unable to reproduce the failures. Here's a branch on a dummy app that shows the provided tests passing: https://github.com/gmcgibbon/rails-issues/tree/37270. It looks like @halo's test should pass on `6-0-stable` too.

As for this still being broken in `6.0`, we can't retroactively patch an already released version of Rails. We can however backport and make sure this fix gets included in `6.0.1`.

@searls can you provide another failing test case if things still aren't working?

**Comment by searls** [comment]
Thanks @gmcgibbon for reopening. Unfortunately I'm heading out the door for vacation won't have access to a machine for about a week, including to tell whether the dummy tests in my original comment pass or fail under 6-0-0-stable@HEAD.

**Comment by beauraF** [comment]
Hi,

In my case, the problem is solved for my `SystemTestCase` by switching to _6-0-stable_.
On the other hand, I still have exactly the same problem with my `ActionDispatch::IntegrationTest`.

**Comment by gmcgibbon** [comment]
Thanks @beauraF. That's precisely what #36283 was trying to sidestep for system test cases. I'm thinking we might have to change the behaviour of `ActiveJob::TestHelper` because it looks like its inclusion in integration tests is intentional. It seems we likely have to delay enabling the test adapter on jobs until just before we assert.

**Comment by beauraF** [comment]
I'm not sure I understood all the ins and outs, but I'd be happy to help you with that @gmcgibbon. It's blocking to get our apps on Rails 6. If there's anything I can do.

**Comment by coorasse** [comment]
I'd like to leave this here, since it could help somebody else. 
In our standard configuration we use rspec and we run system tests with an `inline` adapter, and other tests with `test` adapter.
We had this RSpec configuration:
```
config.around(:each, :inline_jobs) do |example|
  ActiveJob::Base.queue_adapter = :inline
  example.run
  ActiveJob::Base.queue_adapter = :test
end
```

this needs to be changed to:
```
  config.around(:each, :inline_jobs) do |example|
    (ActiveJob::Base.descendants << ActiveJob::Base).each { |a| a.disable_test_adapter }
    ActiveJob::Base.queue_adapter = :inline
    example.run
    (ActiveJob::Base.descendants << ActiveJob::Base).each { |a| a.enable_test_adapter(ActiveJob::QueueAdapters::TestAdapter.new) }
    ActiveJob::Base.queue_adapter = :test
  end
```

**Comment by esaborit4code** [comment]
I had a similar problem with RSpec and request tests, I could workaround that with a little adaption of the code from @coorasse's response above.

```ruby
# spec/support/inline_active_job.rb

RSpec.configure do |config|
  config.around(:each, type: :request) do |example|
    original_queue_adapter = ActiveJob::Base.queue_adapter
    descendants = ActiveJob::Base.descendants + [ActiveJob::Base]

    ActiveJob::Base.queue_adapter = :inline
    descendants.each(&:disable_test_adapter)
    example.run
    descendants.each { |a| a.enable_test_adapter(ActiveJob::QueueAdapters::TestAdapter.new) }
    ActiveJob::Base.queue_adapter = original_queue_adapter
  end
end
```

**Comment by benfb** [comment]
This issue showed up for us because we use the `:sidekiq` adapter even in our test environment. We were testing Sidekiq queue length in two of our tests, one which was a request spec, and the other just a model spec. The model spec had the correct adapter, but the request spec did not. Just providing more context for anyone else who comes across this issue! Adding `(ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter)` to `config.before(:suite)` in our `spec_helper.rb` fixed it.

**Comment by jrochkind** [comment]
Still having what I believe is the same problem on Rails 6.0.1, so alas it didn't get fixed in a 6.0.1 release. 

I'm quite confused about what's going on -- it's like setting `ActiveJob::Base.queue_adapter = something` has no effect in tests for some reason, the tests reset it to some other value. I am trying to set it in a test-by-test basis, to have different tests use different adapters, similar to 
 [@coorassse's use case](https://github.com/rails/rails/issues/37270#issuecomment-546951497) 

I am trying the various workarounds suggested involving `disable_test_adapter` and `enable_test_adapter(ActiveJob::QueueAdapters::TestAdapter.new)` (which I don't fully understand), but haven't gotten any to work yet.

**Comment by jrochkind** [comment]
OK, with my rspec setup, in an rspec/rails `:system` test, what's happening is _first_ the rspec `before` block runs. 

THEN this Rails ActiveJob test_helper before_setup block runs: https://github.com/rails/rails/blob/master/activejob/lib/active_job/test_helper.rb#L36-L46

That before_setup calls `enable_test_adapter` with an argument being the method `queue_adapter_for_test`.  But `queue_adapter_for_test` is hard-coded to  ActiveJob::QueueAdapters::TestAdapter.new https://github.com/rails/rails/blob/master/activejob/lib/active_job/test_helper.rb#L63-L65

I am not sure if this is a bad interaction between rspec-rails and rails that needs to be fixed on rspec side... but I also don't understand how this architecture allows you to change the test ActiveJob adapter at all, it seems to have a `before_setup` that is hard-coded to insist on the Test adapter (ie :test). 

I don't understand what the intent/purpose of that code in activejob/lib/active_job/test_helper.rb is, it's very confusing.

**Comment by searls** [comment]
FWIW, upgrading to 6.0.1 appears to have resolved the particular failure in my test suite that prompted my originally opening the issue. I can't speak to anyone else running into yet related problems, though!

**Comment by jrochkind** [comment]
I am definitely still having a problem in Rails 6.0.1; guess I have to spend more time isolating/reproducing it to figure out exactly what it is, since apparenlty not exactly the same as everyone else. 

It is made more confusing for me by using rspec, which has a whole additional layer on top, especially for integration/system tests. 

But I'm confused about Rails test typology to begin with even before rspec -- there are "System" and "Integration" tests? I don't understand the difference or if one of those is considered "legacy" and discouraged at this point. 

This is relevant because I believe the problem I have is only reproducing with "System" tests, NOT with "Integration" tests or other kinds of tests.  Very confusing.

**Comment by jrochkind** [comment]
It's possible my problem is rspec-rails-specific.  Perhaps the rspec-rails equivalent of this issue, which possibly is already fixed in pure Rails. 

I'm still trying to get to the bottom and wrap my head around what's going on, I'm not certain. I can't myself provide a reproduction without rspec-rails being involved, anyone else?

**Comment by Kevinrob** [comment]
We have the same issue with Rails `6.0.2.1`.
Workaround: adding in `setup`:
```ruby
(ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter)
```
(thanks @searls ❤️ )

**Comment by gaffneyc** [comment]
Still seeing this issue with 6.0.2.2 and [current master](https://github.com/rails/rails/commit/d56d2e740612717334840b0d0ea979e1ca7cf5b1). I've created a [small Rails app to reproduce](https://github.com/gaffneyc/rails_email_deliveries) the problem [specifically this test case](https://github.com/gaffneyc/rails_email_deliveries/blob/master/spec/integration/mailer_spec.rb). We're mainly seeing it as an issue with integration tests. I've run a bisect from Rails 5.2 which showed [this commit](https://github.com/gaffneyc/rails_email_deliveries/blob/master/spec/integration/mailer_spec.rb) as introducing the breakage.

It looks like the issue was found in development but [patched over with this commit](https://github.com/rails/rails/commit/96218b45680e210fcda647401cf4feeb60bfe41b) that fixes the issue in some cases (and only if using `assert_emails`).

To run the example:
```
git clone git@github.com:gaffneyc/rails_email_deliveries.git
cd rails_email_deliveries
bundle
rspec spec/integration/mailer_spec.rb
```

@gmcgibbon Sorry for the ping but is it possible to reopen this issue?

**Comment by benoittgt** [comment]
We added a patch on RSpec to do that **on system spec** but maybe we could handle it for integration spec too.

See https://github.com/rspec/rspec-rails/pull/2242 and https://relishapp.com/rspec/rspec-rails/v/4-0/docs/system-specs/system-spec#the-activejob-queue-adapter-can-be-changed

Could you open an issue?

**Comment by gaffneyc** [comment]
@benoittgt Happy to

**Comment by benoittgt** [comment]
I don't think this issue should be closed. On RSpec Rails has [Jon mentionned](https://github.com/rspec/rspec-rails/issues/2311#issuecomment-633719364)

>we can't really do much without completely "undo'ing" their (rails) work

**Comment by andersonreinkordt** [comment]
FWIW, still seeing this issue on Rails 6.0.3. I will add more details when I have a chance.

**Comment by gaffneyc** [comment]
@gmcgibbon Could we get this issue reopened?

**Comment by nethad** [comment]
I'm not sure how to translate this to RSpec, but I've found that using

```ruby
assert_emails 1 do
  # code under test
end
```

helps here instead of using `ActionMailer::Base.deliveries.size`. I found the solution via https://stackoverflow.com/a/57866436/303626

**Comment by nachbar** [comment]
Thanks, @nethad.  I can confirm that assert_emails works with RSpec, and further that you can still check `ActionMailer::Base.deliveries.size` because, when wrapped in assert_emails, the emails get processed from the `ActiveJob::Base.queue_adapter.enqueued_jobs` array to `ActionMailer::Base.deliveries`

**Comment by yevhene** [comment]
Rails still force TestAdapter to be used. None of the solution didn't help. Only spreading `perform_enqueued_jobs` helped me.

**Comment by yevhene** [comment]
This helped to me:
```  
config.before(:each) do
  (ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter)
  ActiveJob::Base.queue_adapter = :sidekiq
end
```

**Comment by jsharpe-ctm** [comment]
Verified this is still an issue on 6.0.3.5.  Also verified @yevhene's 'fix' works for me in a minitest/resque env.  This works for me:

```
class ActionDispatch::IntegrationTest
  def setup
    (ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter)
  end
end
```

@georgeclaghorn would it be possible to update the milestones to target some 6.1.next?

**Comment by aiomaster** [comment]
The only thing that worked for me was something like:

```ruby
def setup
  queue_adapter_changed_jobs.each(&:disable_test_adapter)
end
```
cause the line suggested in this thread
```ruby
(ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter)
```
broke other tests, that uses the test adapter and were executed after my test with the real queue adapter. I don't really understand why it works that way, but for me it does.

**Comment by ansonhoyt** [comment]
Thanks @aiomaster for the tip, as my later tests broke too. I couldn't call `queue_adapter_changed_jobs` directly since `include ActiveJob::TestHelper` seems to force the `:test` adapter, and I'm overriding with `:delayed_job`. So, I updated the `descendants = ` line from @esaborit4code's solution with the implementation from [queue_adapter_changed_jobs](https://github.com/rails/rails/blob/6-1-stable/activejob/lib/active_job/test_helper.rb#L694).

Switching adapters from `:async` to `:test` can also break later tests. Any `:async` jobs may still be running while your next test is running, which is tricky to see since the adapter has changed and the current queue may appear empty. I avoided that possibility by calling [shutdown](https://ruby-concurrency.github.io/concurrent-ruby/1.1.4/Concurrent/ThreadPoolExecutor.html#shutdown-instance_method):
```ruby
# spec/support/async_queue_cleanup.rb
RSpec.configure do |config|
  # Ensure no jobs are executed during the next context or example group.
  # Avoided `after(:each)` in case waiting for shutdown more often slowed the test suite.
  config.after(:all) do |example|
    ActiveJob::Base.queue_adapter.shutdown if ActiveJob::Base.queue_adapter.is_a? ActiveJob::QueueAdapters::AsyncAdapter
  end
end

```

**Comment by rept** [comment]
> I could not even move the patch to the RSpec `config.around` block, it works only right here, in the test.
> 
> I started noticing in my test logs that, in Rails 6, the job is only enqueued but not run. (And when I wrapped the test in `perform_enqueued_jobs do`, the job ran, but the `provider_job_id` was not matching from when the job was enqueued compared to when the websocket channel is set up.)

@halo I have the exact same issue in Rails 6.1.3. and only including this code:

`    (ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter)`
in the actual test itself solves it, exactly like you describe.  I use Capybara too, maybe related.

**Comment by jrochkind** [comment]
Oddly this DOES now appear to be working fine for me with rails 6.1.4.1,  rspec-rails 5.0.2, and rspec 3.10.0. 

I'm not sure if everyone else still having the problem are using rspec, or if some are having it with straight Rails out of the box testing. 

I think it's been working for me for a while, not just with those recent versions. 

I think it's been working for me ever since this PR landed in rspec-rails: https://github.com/rspec/rspec-rails/pull/2242

Although I'm having trouble identifying exactly what version of rspec-rails that is, when I try to get github to tell me it reports "This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository" (maybe a rebase involved?). 

If you are using rspec... it may depend on what type of rspec test you are using? That PR fixes things for "system" tests -- are those of you still having the problem using "feature" tests, and having the problem there? (Or in other types? Controller? etc?).  It's possible switching from "system" to "feature" type tests might help for you?

**Comment by rept** [comment]
@jrochkind just updated rspec-rails to 5.0.2 from a version < 4.1 but unfortunately same problem still.  I'm using **type: :request** though.

Putting this line in the actual test is the only thing that fixes it:
`(ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter)`

**Comment by halo** [comment]
@jrochkind Thank you for the news. I can confirm that this bug is resolved and that it's working now without any patches. It works with both `type: :system` and `type: :request`.

I did not change my tests nor any of the test setup. But I've been running `bundle update` regularly. Currently I'm on these version (I can't tell at which version the bug disappeared though):

```
rails 6.1.4.1
rspec 3.10.0
rspec-rails 5.0.2
capybara 3.35.3
```

If there are a more people reporting that the bug disappeared, I'm hopeful this issue can simply be closed.

**Comment by jrochkind** [comment]
It is mysterious and frustrating!  

It definitely matters what kind of test it was. So interstingly, there was a previous bug in Rails related to this that only applied to *system* tests, that was fixed in: https://github.com/rails/rails/pull/36283

However, that PR makes system tests "Inherit from ActiveSupport::TestCase instead of ActionDispatch::IntegrationTest" to fix the problem -- which implies the problem might still be there in IntegrationTests, or perhaps Integration tests are intentionally not designed to allow queue_adapter setting? 

You say you are using rspec requests tests... but guess what, rspec [says]:(https://relishapp.com/rspec/rspec-rails/docs/request-specs/request-spec) "Request specs provide a thin wrapper around Rails' integration test",

There are two things to tease apart here:

1. Separating rspec from rails -- this is the rails issue tracker, if it's an rspec-specific problem, or if it hasn't been confirmed which it is, it's especially less likely to result in anything here
1. What *type* of test are people having the problem in -- this may be test-type specific

So, while a pain, if I were you and had time/energy and wanted to try to help get somewhere here, I'd try making a fresh empty Rails app that has nothing in it, and seeing if I can reproduce the problem without usingRspec, using a [Rails integration test specifically](https://guides.rubyonrails.org/testing.html#integration-testing).

Then report that with the reproduction (if you can do it as a [self-contained test-case](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#create-an-executable-test-case), even better) -- either here, or make a new issue with more specific scope and concise problem report. 

Sorry, I know a bug like this is annoying.

**Comment by rept** [comment]
@jrochkind I tried making it as self-contained-test but could get that to work. However I think I do have it reproduced with new empty rails app (minimal).

Just to make sure we're on the same page:

If I have 

`config.active_job.queue_adapter = :inline`
in my test.rb

Then the job should run inline right?

require "test_helper"

```
class SomeTest < ActionDispatch::IntegrationTest
  test "did job run" do
    p ActiveJob::Base.queue_adapter
    SomeJob.perform_later
    assert_equal true, SomeJob.job_ran
  end
end
```

This is failing for me.

Repo here: https://github.com/rept/issue_37270

**Comment by rgaufman** [comment]
Thank you for the above, this is indeed very frustrating and took us ages to figure out, this fixes it <

```
ActiveJob::Base.descendants << ActiveJob::Base).each(&:disable_test_adapter
```

**Comment by cdan-youdo** [comment]
My workaround was to wrap the code inside a
```
perform_enqueued_jobs do
....
end
```

This way, any jobs that are enqueued inside the block, will be run automatically.

**Comment by ghiculescu** [comment]
Hi everyone! I'd love some feedback on this: https://github.com/rails/rails/pull/48585

**Comment by dorianmarie** [comment]
rails 7.1 and this is still an issue

**Comment by jmarsh24** [comment]
+1 for resolving this issue!

**Comment by rgaufman** [comment]
Is this fixed in Rails 8?
